src/device_memory_tracker.cc

```diff
@@ -0,0 +1,288 @@
+       : (buffer))
+
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
+}
+
+}  // namespace
+
+DeviceMemoryTracker::DeviceMemoryTracker()
+{
+  cudaError_t cuerr = cudaGetDeviceCount(&device_cnt_);
+  if ((cuerr == cudaErrorNoDevice) || (cuerr == cudaErrorInsufficientDriver)) {
+    device_cnt_ = 0;
+  } else if (cuerr != cudaSuccess) {
+    throw std::runtime_error("Unexpected failure on getting CUDA device count.");
+  }
+
+  // Use 'cuptiSubscribe' to check if the cupti has been initialized
+  // elsewhere. Due to cupti limitation, there can only be one cupti client
+  // within the process, so in the case of per-backend memory tracking, we
+  // have to make the assumption that the other cupti client is using the same
+  // memory tracker implementation so that the backend may use the cupti
+  // configuration that is external to the backend without issue.
+  auto cupti_res = cuptiSubscribe(&subscriber_, nullptr, nullptr);
+  switch (cupti_res)
+  {
+  case CUPTI_SUCCESS : {
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
+    break;
+  }
+  default:
+    // other error, should propagate and disable memory tracking for the backend
+    throw std::runtime_error("Unexpected failure on configuring CUPTI.");
+  }
+}
+
+int
+DeviceMemoryTracker::CudaDeviceCount()
+{
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  return 0;
+}
+
+bool
+DeviceMemoryTracker::Init()
+{
+  if (tracker_ == nullptr) {
```

Comment:
During local testing - I found this to not be thread safe in the case of calling Init() from a ModelInitialize(). So that if you have a backend responsible for multiple models and it calls Init() - there's a chance that both will call DeviceMemoryTracker(). Only one will succeed in subscribing - but the other one will be destroyed and can lead to confusion. Suggest a mutex

Reply:
`DeviceMemoryTracker::Init()` should be called on `TRITONBACKEND_Initialize()`

Reply:
thanks!